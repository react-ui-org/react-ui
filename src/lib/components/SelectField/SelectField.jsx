import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootSizeClassName } from '../_helpers/getRootSizeClassName';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { resolveContextOrProp } from '../_helpers/resolveContextOrProp';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import withForwardedRef from '../withForwardedRef';
import styles from './SelectField.scss';

export const SelectField = ({
  disabled,
  forwardedRef,
  fullWidth,
  helpText,
  id,
  isLabelVisible,
  label,
  layout,
  options,
  required,
  size,
  validationState,
  validationText,
  value,
  variant,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        context && styles.isRootInFormLayout,
        resolveContextOrProp(context && context.layout, layout) === 'horizontal'
          ? styles.rootLayoutHorizontal
          : styles.rootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootSizeClassName(size, styles),
        getRootValidationStateClassName(validationState, styles),
        variant === 'filled' ? styles.rootVariantFilled : styles.rootVariantOutline,
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div
        className={classNames(
          styles.label,
          !isLabelVisible && styles.isLabelHidden,
        )}
        id={id && `${id}__labelText`}
      >
        {label}
      </div>
      <div className={styles.field}>
        <div className={styles.inputContainer}>
          <select
            {...transferProps(restProps)}
            className={styles.input}
            disabled={disabled}
            id={id}
            ref={forwardedRef}
            required={required}
            value={value}
          >
            {
              options.map((option) => (
                <option
                  disabled={option.disabled}
                  id={id && `${id}__item__${option.value}`}
                  key={option.value}
                  value={option.value}
                >
                  {option.label}
                </option>
              ))
            }
          </select>
          <div className={styles.caret}>
            <span className={styles.caretIcon} />
          </div>
          {variant === 'filled' && (
            <div className={styles.bottomLine} />
          )}
        </div>
        {helpText && (
          <div
            className={styles.helpText}
            id={id && `${id}__helpText`}
          >
            {helpText}
          </div>
        )}
        {validationText && (
          <div
            className={styles.validationText}
            id={id && `${id}__validationText`}
          >
            {validationText}
          </div>
        )}
      </div>
    </label>
  );
};

SelectField.defaultProps = {
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helpText: null,
  id: undefined,
  isLabelVisible: true,
  layout: 'vertical',
  required: false,
  size: 'medium',
  validationState: null,
  validationText: null,
  value: undefined,
  variant: 'outline',
};

SelectField.propTypes = {
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `select` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the input HTML element.
   *
   * Also serves as a prefix for important inner elements:
   * * `<ID>__label`
   * * `<ID>__labelText`,
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   *
   * and of individual options:
   * * `<ID>__item__<VALUE>`
   */
  id: PropTypes.string,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Select field label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Layout of the field.
   *
   * Ignored if the component is rendered within `FormLayout` component
   * as the value is inherited in such case.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * Set of options to be chosen from.
   */
  options: PropTypes.arrayOf(PropTypes.shape({
    disabled: PropTypes.bool,
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  })).isRequired,
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Size of the field.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
  /**
   * Value of the input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Design variant of the field, further customizable with CSS custom properties.
   */
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export const SelectFieldWithContext = withForwardedRef(withProviderContext(SelectField, 'SelectField'));

export default SelectFieldWithContext;
