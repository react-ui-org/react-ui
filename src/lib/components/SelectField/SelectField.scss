@use "../../styles/tools/form-fields/box-field-elements";
@use "../../styles/tools/form-fields/box-field-layout";
@use "../../styles/tools/form-fields/box-field-sizes";
@use "../../styles/tools/form-fields/foundation";
@use "../../styles/tools/form-fields/variants";
@use "../../styles/tools/accessibility";

// Foundation
.root {
    @include foundation.root();
}

.label {
    @include foundation.label();
}

.inputContainer {
    @include box-field-elements.input-container();
}

.input {
    @include box-field-elements.input();
    @include box-field-elements.input-select();
    @include box-field-elements.input-select-option();
}

.caret {
    @include box-field-elements.caret();
}

.caretIcon {
    @include box-field-elements.caret-icon();
}

.bottomLine {
    @include box-field-elements.bottom-line();
}

.helpText,
.validationText {
    @include foundation.help-text();
}

.isRootRequired .label {
    @include foundation.label-required();
}

// Variants
.rootVariantFilled {
    @include variants.visual(box, $variant: filled, $has-caret: true);
}

.rootVariantOutline {
    @include variants.visual(box, $variant: outline, $has-caret: true);
}

// States
.isRootStateInvalid {
    @include variants.validation(invalid);
}

.isRootStateValid {
    @include variants.validation(valid);
}

.isRootStateWarning {
    @include variants.validation(warning);
}

// Invisible label
.isLabelHidden {
    @include accessibility.hide-text();
}

// Layouts
.rootLayoutVertical,
.rootLayoutHorizontal {
    @include box-field-layout.vertical();
}

.rootLayoutHorizontal {
    @include box-field-layout.horizontal();
}

.isRootFullWidth {
    @include box-field-layout.full-width();
}

.isRootInFormLayout {
    @include box-field-layout.in-form-layout();
}

// Sizes
.rootSizeSmall {
    @include box-field-sizes.size(small);
}

.rootSizeMedium {
    @include box-field-sizes.size(medium);
}

.rootSizeLarge {
    @include box-field-sizes.size(large);
}
