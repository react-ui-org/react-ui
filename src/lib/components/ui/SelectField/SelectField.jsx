import PropTypes from 'prop-types';
import React from 'react';
import getRootSizeClassName from '../../../helpers/getRootSizeClassName';
import getRootValidationStateClassName from '../../../helpers/getRootValidationStateClassName';
import { withProviderContext } from '../../../provider';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './SelectField.scss';

export const SelectField = ({
  changeHandler,
  disabled,
  forwardedRef,
  fullWidth,
  helpText,
  id,
  inFormLayout,
  isLabelVisible,
  label,
  layout,
  options,
  required,
  size,
  validationState,
  validationText,
  value,
  variant,
  ...restProps
}) => (
  <label
    className={[
      styles.root,
      fullWidth ? styles.isRootFullWidth : '',
      inFormLayout ? styles.isRootInFormLayout : '',
      layout === 'horizontal' ? styles.rootLayoutHorizontal : styles.rootLayoutVertical,
      disabled ? styles.isRootDisabled : '',
      required ? styles.isRootRequired : '',
      getRootSizeClassName(size, styles),
      getRootValidationStateClassName(validationState, styles),
      variant === 'filled' ? styles.rootVariantFilled : styles.rootVariantOutline,
    ].join(' ')}
    htmlFor={id}
    id={id && `${id}__label`}
  >
    <div
      className={[
        styles.label,
        isLabelVisible ? '' : styles.isLabelHidden,
      ].join(' ')}
      id={id && `${id}__labelText`}
    >
      {label}
    </div>
    <div className={styles.field}>
      <div className={styles.inputContainer}>
        <select
          {...transferProps(restProps)}
          className={styles.input}
          disabled={disabled}
          id={id}
          onChange={changeHandler}
          ref={forwardedRef}
          required={required}
          value={value}
        >
          {
            options.map((option) => (
              <option
                disabled={option.disabled}
                id={id && `${id}__item__${option.value}`}
                key={option.value}
                value={option.value}
              >
                {option.label}
              </option>
            ))
          }
        </select>
        <div className={styles.caret}>
          <span className={styles.caretIcon} />
        </div>
        {variant === 'filled' && (
          <div className={styles.bottomLine} />
        )}
      </div>
      {helpText && (
        <div
          className={styles.helpText}
          id={id && `${id}__helpText`}
        >
          {helpText}
        </div>
      )}
      {validationText && (
        <div
          className={styles.validationText}
          id={id && `${id}__validationText`}
        >
          {validationText}
        </div>
      )}
    </div>
  </label>
);

SelectField.defaultProps = {
  changeHandler: null,
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helpText: null,
  id: undefined,
  inFormLayout: false,
  isLabelVisible: true,
  layout: 'vertical',
  required: false,
  size: 'medium',
  validationState: null,
  validationText: null,
  value: undefined,
  variant: 'outline',
};

SelectField.propTypes = {
  /**
   * Function to call when the input has changed.
   */
  changeHandler: PropTypes.func,
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `select` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the input HTML element. It also serves as a prefix for important inner elements:
   * `<ID>__label`, `<ID>__labelText`, `<ID>__helpText`, `<ID>__validationText`, and all options:
   * `<ID>__item__<VALUE>`.
   */
  id: PropTypes.string,
  /**
   * Treat the field differently when it's inside a FormLayout. Do not set manually!
   */
  inFormLayout: PropTypes.bool,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Select field label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Layout of the field.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * Set of options to be chosen from.
   */
  options: PropTypes.arrayOf(PropTypes.shape({
    disabled: PropTypes.bool,
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  })).isRequired,
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Size of the field.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
  /**
   * Value of the input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Design variant of the field, further customizable with CSS custom properties.
   */
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export const SelectFieldWithContext = withForwardedRef(withProviderContext(SelectField, 'SelectField'));

export default SelectFieldWithContext;
