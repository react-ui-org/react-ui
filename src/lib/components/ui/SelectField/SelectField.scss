@use '../../../styles/tools/accessibility';
@use '../../../styles/tools/forms/foundation';
@use '../../../styles/tools/forms/layouts';
@use '../../../styles/tools/forms/relationships';
@use '../../../styles/tools/forms/sizes';
@use '../../../styles/tools/forms/variants';

// Foundation
.root {
  @include foundation.field();
  @include relationships.field-horizontal-neighbor();
}

.label {
  @include foundation.field-label();
}

.inputContainer {
  @include foundation.field-input-container();
}

.input {
  @include foundation.field-input();
  @include foundation.field-input-select();
  @include foundation.field-input-select-option();
}

.caret {
  @include foundation.field-caret();
}

.caretIcon {
  @include foundation.field-caret-icon();
}

.bottomLine {
  @include foundation.field-bottom-line();
}

.helpText,
.validationText {
  @include foundation.field-help-text();
}

.isRootRequired .label {
  @include foundation.field-label-required();
}

// Variants
.rootVariantFilled {
  @include variants.field-variant(filled, $has-caret: true);
}

.rootVariantOutline {
  @include variants.field-variant(outline, $has-caret: true);
}

// States
.isRootStateInvalid {
  @include variants.field-variant(invalid, $has-caret: true);
}

.isRootStateValid {
  @include variants.field-variant(valid, $has-caret: true);
}

.isRootStateWarning {
  @include variants.field-variant(warning, $has-caret: true);
}

// Invisible label
.isLabelHidden {
  @include accessibility.hide-text();
}

// Layouts
.rootLayoutHorizontal {
  @include layouts.field-layout-horizontal();
}

.rootLayoutVertical {
  @include layouts.field-layout-vertical();
}

.isRootFullWidth {
  @include layouts.field-full-width();
  @include relationships.field-vertical-neighbor();
}

.isRootInFormLayout {
  @include layouts.field-in-form-layout();
}

// Sizes
.rootSizeSmall {
  @include sizes.field-size(small);
}

.rootSizeMedium {
  @include sizes.field-size(medium);
}

.rootSizeLarge {
  @include sizes.field-size(large);
}
