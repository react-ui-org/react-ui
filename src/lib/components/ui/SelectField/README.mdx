---
name: SelectField
menu: 'UI'
route: /components/ui/select-field
---

# SelectField

SelectField allows users to select one option from a set.

import {
  Playground,
  Props,
} from 'docz'
import { SelectField } from './SelectField'

## Basic Usage

To implement the SelectField component, you need to import it first:

```js
import { SelectField } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    return (
      <SelectField
        changeHandler={(e) => setFruit(e.target.value)}
        id="my-select-field"
        label="Your favourite fruit"
        options={[
          {
            label: 'Apple',
            value: 'apple',
          },
          {
            label: 'Banana',
            value: 'banana',
          },
          {
            label: 'Grapefruit',
            value: 'grapefruit',
          },
        ]}
        value={fruit}
      />
    );
  }}
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use SelectField for **many options**. For sets of just a few options
  (say 3 at maximum) consider using the [Radio](/components/ui/radio) component.
  This will help keep your UI clean and uncluttered and prevent your users from
  being overwhelmed by too many options.

- **Don't use for boolean** (true/false) selection or to toggle things on and
  off. [CheckboxField](/components/ui/checkbox-field) and
  [Toggle](/components/ui/toggle) are more suitable for such cases.

- Use **short and descriptive labels**, ideally nouns rather than seemingly
  polite phrases like _Please select your favourite fruit_. Short labels will
  help your users accomplish their task faster.

- Only make the SelectField's label invisible when there is **another visual
  clue** to guide users through filling the input. Using the first option as
  label is not recommended either â€” it disappears once user makes their choice.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Use **clear, calm error messages** when there's a problem with what they
  entered.

- In case of a large amount of options, consider **grouping related items
  together** by nesting them (implements the `optgroup` HTML element).

ðŸ‘‰ We use the **native `select`** HTML element to improve user experience on
mobile devices by using the native select of the platform.

## Design Variants

To satisfy the design requirements of your project, all input fields in React UI
come in two design variants to choose from: outline and filled. Both can be
further [customized](#theming) with CSS custom properties.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline"
          label="Your favourite fruit"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled"
          label="Your favourite fruit"
          options={options}
          variant="filled"
          value={fruit}
        />
      </>
    );
  }}
</Playground>

## Sizes

Aside from the default (medium) size, two additional sizes are available: small
and large.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-small"
          label="Your favourite fruit"
          options={options}
          size="small"
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-medium"
          label="Your favourite fruit"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-large"
          label="Your favourite fruit"
          options={options}
          size="large"
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-small"
          label="Your favourite fruit"
          options={options}
          size="small"
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-medium"
          label="Your favourite fruit"
          options={options}
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-large"
          label="Your favourite fruit"
          options={options}
          size="large"
          value={fruit}
          variant="filled"
        />
      </>
    );
  }}
</Playground>

Full-width fields span the full width of a parent:

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          fullWidth
          id="my-select-field-outline-full-width"
          label="Your favourite fruit"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          fullWidth
          id="my-select-field-filled-full-width"
          label="Your favourite fruit"
          options={options}
          value={fruit}
          variant="filled"
        />
      </>
    );
  }}
</Playground>

## Invisible Label

While it may be acceptable for login screens with just a few fields or for other
simple forms, it's dangerous to hide labels from users in most cases. Keep in
mind you should **provide another visual clue** so users know what to fill into
the input.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-hidden-label"
          isLabelVisible={false}
          label="Your favourite fruit"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-hidden-label"
          isLabelVisible={false}
          label="Your favourite fruit"
          options={options}
          value={fruit}
          variant="filled"
        />
      </>
    );
  }}
</Playground>

## Horizontal Layout

The default vertical layout is very easy to use and work with. However, there
are situations where horizontal layout suits better â€” and that's why React UI
supports this kind of layout as well.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-horizontal"
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-horizontal"
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          fullWidth
          id="my-select-field-outline-horizontal-full-width"
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          fullWidth
          id="my-select-field-filled-horizontal-full-width"
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          fullWidth
          id="my-select-field-outline-horizontal-full-width-invisible-label"
          isLabelVisible={false}
          label="Your favourite fruit"
          layout="horizontal"
          value={fruit}
          options={options}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          fullWidth
          id="my-select-field-filled-horizontal-full-width-invisible-label"
          isLabelVisible={false}
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
          variant="filled"
        />
      </>
    );
  }}
</Playground>

## Help Text

You may provide an additional help text to clarify how the input should be
filled.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          helpText="Choose one or more kinds of fruit to feel happy."
          id="my-select-field-outline-help-text"
          label="Your favourite fruit"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          helpText="Choose one or more kinds of fruit to feel happy."
          id="my-select-field-filled-help-text"
          label="Your favourite fruit"
          options={options}
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          helpText="Choose one or more kinds of fruit to feel happy."
          id="my-select-field-outline-help-text-horizontal"
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          helpText="Choose one or more kinds of fruit to feel happy."
          id="my-select-field-filled-help-text-horizontal"
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          fullWidth
          helpText="Choose one or more kinds of fruit to feel happy."
          id="my-select-field-outline-help-text-horizontal-full-width"
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          fullWidth
          helpText="Choose one or more kinds of fruit to feel happy."
          id="my-select-field-filled-help-text-horizontal-full-width"
          label="Your favourite fruit"
          layout="horizontal"
          options={options}
          value={fruit}
          variant="filled"
        />
      </>
    );
  }}
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide a validation message for states other than valid** so
users know what happened and what action they should take or what options they
have.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-success"
          label="Your favourite fruit"
          options={options}
          required
          validationState="valid"
          validationText="Great, they're in stock!"
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-warning"
          label="Your favourite fruit"
          options={options}
          required
          validationState="warning"
          validationText="Oh, really?"
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-error"
          label="Your favourite fruit"
          options={options}
          required
          validationState="invalid"
          validationText="You must select at least one kind of fruit."
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-success"
          label="Your favourite fruit"
          options={options}
          required
          validationState="valid"
          validationText="Great, they're in stock!"
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-warning"
          label="Your favourite fruit"
          options={options}
          required
          validationState="warning"
          validationText="Oh, really?"
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-error"
          label="Your favourite fruit"
          options={options}
          required
          value={fruit}
          validationState="invalid"
          validationText="You must select at least one kind of fruit."
          variant="filled"
        />
      </>
    );
  }}
</Playground>

### Disabled State

It's possible to disable just some options or the whole input.

<Playground>
  {() => {
    const [fruit, setFruit] = React.useState('apple');
    const options = [
      {
        label: 'Apple',
        value: 'apple',
      },
      {
        disabled: true,
        label: 'Banana',
        value: 'banana',
      },
      {
        label: 'Grapefruit',
        value: 'grapefruit',
      },
    ];
    return (
      <>
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-outline-disabled-option"
          label="Your favourite fruit"
          options={options}
          value={fruit}
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          id="my-select-field-filled-disabled-option"
          label="Your favourite fruit"
          options={options}
          value={fruit}
          variant="filled"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          disabled
          id="my-select-field-outline-disabled"
          label="Your favourite fruit"
          options={options}
          value="apple"
        />
        <SelectField
          changeHandler={(e) => setFruit(e.target.value)}
          disabled
          id="my-select-field-filled-disabled"
          label="Your favourite fruit"
          options={options}
          value="apple"
          variant="filled"
        />
      </>
    );
  }}
</Playground>

## API

In addition to the options below, you can add any custom attributes that do not
interfere with the API, and they will be passed to the `select` HTML element.
This is useful mainly to improve component's accessibility.

<Props table of={SelectField} />

## Theming

Head to [Forms Theming](/customize/theming/forms) to see shared form theming
options. On top of that, the following options are available for SelectField.

| Custom Property                                      | Description                                                  |
|------------------------------------------------------|--------------------------------------------------------------|
| `--rui-FormField--box--select__caret__border-style`  | SelectField arrow border style (eg. `solid`)                 |
| `--rui-FormField--box--select__caret__background`    | SelectField arrow background (including `url()` or gradient) |
| `--rui-FormField--box--select__option--disabled__color` | Text color of disabled SelectField options                |
