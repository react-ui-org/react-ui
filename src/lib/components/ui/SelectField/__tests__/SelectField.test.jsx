import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { SelectField } from '../SelectField';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<SelectField
      id="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly only one option', () => {
    const tree = shallow(<SelectField
      id="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<SelectField
      disabled
      fullWidth
      helpText="some help"
      id="test"
      inFormLayout
      isLabelVisible={false}
      label="label"
      layout="horizontal"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          disabled: true,
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      required
      size="large"
      validationState="invalid"
      value="ch1"
      variant="filled"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls changeHandler() on changing selected option', () => {
    const spy = sinon.spy();
    const component = mount(<SelectField
      changeHandler={spy}
      id="id"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      value="ch1"
    />);

    component
      .find('select')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'ch2' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
