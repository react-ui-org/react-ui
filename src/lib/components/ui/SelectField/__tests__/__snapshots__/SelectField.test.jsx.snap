// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<label
  className="root   rootLayoutVertical  rootSizeMedium  rootVariantOutline"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label "
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <select
        className="input"
        disabled={false}
        id="test"
        onChange={null}
        required={false}
      >
        <option
          id="test__item__ch1"
          key="ch1"
          value="ch1"
        >
          choice 1
        </option>
        <option
          id="test__item__ch2"
          key="ch2"
          value="ch2"
        >
          choice 2
        </option>
      </select>
      <div
        className="caret"
      >
        <span
          className="caretIcon"
        />
      </div>
    </div>
  </div>
</label>
`;

exports[`rendering renders correctly only one option 1`] = `
<label
  className="root   rootLayoutVertical  rootSizeMedium  rootVariantOutline"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label "
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <select
        className="input"
        disabled={false}
        id="test"
        onChange={null}
        required={false}
      >
        <option
          id="test__item__ch1"
          key="ch1"
          value="ch1"
        >
          choice 1
        </option>
      </select>
      <div
        className="caret"
      >
        <span
          className="caretIcon"
        />
      </div>
    </div>
  </div>
</label>
`;

exports[`rendering renders correctly with all props 1`] = `
<label
  className="root isRootFullWidth isRootInFormLayout rootLayoutHorizontal isRootRequired rootSizeLarge isRootStateInvalid rootVariantFilled"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label isLabelHidden"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="field"
  >
    <div
      className="inputContainer"
    >
      <select
        className="input"
        disabled={true}
        id="test"
        onChange={null}
        required={true}
        value="ch1"
      >
        <option
          id="test__item__ch1"
          key="ch1"
          value="ch1"
        >
          choice 1
        </option>
        <option
          disabled={true}
          id="test__item__ch2"
          key="ch2"
          value="ch2"
        >
          choice 2
        </option>
      </select>
      <div
        className="caret"
      >
        <span
          className="caretIcon"
        />
      </div>
      <div
        className="bottomLine"
      />
    </div>
    <div
      className="helpText"
      id="test__helpText"
    >
      some help
    </div>
  </div>
</label>
`;
